### DISCUSSION TITLE
Should `mise i go` set GOBIN?

### DISCUSSION THREAD

**Discussion by powerman** [jdx/mise#2929] [General]
Current behaviour:

```
$ go env | grep GOBIN
GOBIN='/home/powerman/.local/share/mise/installs/go/1.23.2/bin'
```

I think it makes more sense to change this and do not set GOBIN, to have this result:

```
$ go env | grep GOBIN
GOBIN=''
```

Documentation about GOBIN:

```
$ go help environment | grep -A 1 GOBIN
	GOBIN
		The directory where 'go install' will install a command.
```

More [details](https://pkg.go.dev/cmd/go#hdr-Compile_and_install_packages_and_dependencies):

> Executables are installed in the directory named by the `GOBIN` environment variable, which defaults to `$GOPATH/bin` or `$HOME/go/bin` if the `GOPATH` environment variable is not set.

The point is:
- Go compiles an app to a static binary which is not depends on current Go package.
- The binaries installed by `go install` are supposed to be available in user's PATH until user will manually remove them. They shouldn't disappear from PATH just because Mise switch to a different go version or even uninstalls go.
- Users usually expect to install these binaries into `$GOPATH/bin` (because they usually do not set `GOBIN`, but may set `GOPATH`).
- User may execute `go install github.com/some/thing@v1.2.3` at any moment in any directory and expects to have `thing` binary available in his PATH. Installed thing often is not related to current working directory/project, it's more like random `mise i` or `mise exec thing@1.2.3 -- thing`.

**Comment by powerman** [comment]
@jdx I'm pretty sure it's a bug to be fixed. You can test it yourself:

```console
$ go env GOBIN

$ which -a hello
hello not found
$ mkdir mise-gobin && cd mise-gobin && touch mise.toml && mise trust && mise use go@latest
$ go install golang.org/x/example/hello@latest
$ go env GOBIN
/home/powerman/.local/share/mise/installs/go/1.24.3/bin
$ which -a hello
/home/powerman/.local/share/mise/installs/go/1.24.3/bin/hello
$ hello
Hello, world!
$ cd ..
$ hello
zsh: command not found: hello
```

Unlike tools installed by Mise, tools installed by `go install` must not disappears on directory change. It does not matter here how `go` itself was installed and which version of `go` was used to install some tool.

**Comment by liamdawson** [comment]
Setting GOBIN has side effects on other utilities, too. See https://github.com/nao1215/gup/issues/206

I'm of the opinion that mise shouldn't be setting GOBIN, thankfully it's configurable via `MISE_GO_SET_GOBIN=false`.
